hack/generator/spec/versioning.md

```diff
@@ -0,0 +1,201 @@
+
+## Proposed Solution
+
+In summary (details below):
+
+* Base the storage version on the latest available API version for each resource type
+* Augment the storage version with a name/value property bag for storing additional information
+* Generate the required `ConvertTo()` and `ConvertFrom()` methods automatically
+* Allow configuration to support common changes from version to version
+* Define extension points for hand written conversions
+
+### Use the latest available API Version
+
+As a starting point for defining the storage schema for each resource type, we will use the latest available version of the matching ARM API. This ensures maximum compatibility with that API when it is used, giving a good starting point for new adoption of the service operator.
+
+If a resource type has been dropped from the ARM API, we will still generate a storage schema for it based on the last ARM API version where it existed; this ensures backward compatibility with existing service operator deployments.
+
+### Augment storage with a name/value property bag
+
+To facilitate easy and safe storage of additional configuration in storage, we'll include a name/value property bag for storage, along with helper methods that allow for strongly typed access to values stored within.
+
+Sequestering the information away within the storage schema is more robust than using separate annotations as they are less subject to arbitary modification by users. This allows us to largely avoid situations where well meaning (but uninformed) consumers of the service operator innocently make changes that result in the operator becoming failing. We particularly want to avoid this failure mode because recovery will be difficult - restoration of the modified/deleted information may be impractical or impossible.
+
+---
+
+****Outstanding issue:** Where do the helper functions for the name/value property bag reside?**
+
+We could generate those anew for each type, making them conveniently available and avoiding the need for our generated types to depend on anything more than the standard Go library Alternatively, we could define the property bag as a new struct, wrapping up storage & helper methods into a reusable form.
+
+---
+
+### Generate conversion methods
+
+Each of the structs generated for ARM API will have the normal `ConvertTo()` and `ConvertFrom()` methods generated automatically, mapping like-for-like properties between the hub version and the current type.
+
+* When the name and type of properties match exactly, conversion in both directions will be simple assignments.
+* Where the name matches but the type does not, the property will be skipped by default.
```

Comment:
What do you mean by "skipped"? Not dealt with at all? Is this where it would rely on us explicitly informing the generator how to handle this case? Is this where the configuration comes in?

Reply:
Yeah, skipped is less than ideal because it's silent. I'll capture this as an outstanding issue.